### PR TITLE
Define isleaf

### PR DESCRIPTION
**Pull request by CarloLucibello**
In order to allow for customized data movement for custom types we give the possibility to hook into the data transfer mechanism by implementing
```julia
MLDataDevices.isleaf(x::MyType) = true
Adapt.adapt_structure(device::CUDADevice, x::MyType) = ...  
```

See
https://github.com/FluxML/Flux.jl/pull/2498#issuecomment-2409050719
and the discussion in that PR.

TODO
- [x] tests
- [x] docs

### CODE REVIEW

**Comment by CarloLucibello** [comment]
@avik-pal could you give me feedback if I should finish this PR or if you prefer another approach?

File: src/MLDataDevices.jl [hunk 1]
```diff
@@ -25,4 +25,7 @@ export get_device, get_device_type
 
 export DeviceIterator
 
+### uncomment below when min supported julia version is >=1.11
+# public isleaf 
+
```
Comment (avik-pal) [inline src/MLDataDevices.jl]: Add via `Compat.jl`

File: src/public.jl [hunk 2]
```diff
@@ -370,3 +370,6 @@ for T in (AMDGPUDevice, AMDGPUDevice{Nothing}, CUDADevice,
     CUDADevice{Nothing}, MetalDevice, oneAPIDevice)
     @eval Adapt.adapt_storage(to::$(T), x::AbstractRange) = Adapt.adapt(to, collect(x))
 end
+
+
+isleaf(x) = Functors.isleaf(x)
```
Comment (avik-pal) [inline src/public.jl]: Add a docstring mentioning that this only overrides the behavior of types excluding `AbstractArray` / `Tuple` / `NamedTuple`

**Review by avik-pal — COMMENTED** [review]
overall looks good.

**Comment by CarloLucibello** [comment]
done

File: Project.toml [hunk 3]
```diff
@@ -48,6 +49,7 @@ AMDGPU = "0.9.6, 1"
 Adapt = "4"
 CUDA = "5.2"
 ChainRulesCore = "1.23"
+Compat = "4.16.0"
```
Comment (avik-pal) [inline Project.toml]:
```suggestion
Compat = "4.15"
```

**Review by avik-pal — APPROVED** [review]

**Pull request merged**